thrust/thrust/iterator/transform_iterator.h

```diff
@@ -41,202 +41,236 @@
- *    {
- *      return x * x;
- *    }
- *  };
- *
- *  int main()
- *  {
- *    // initialize a device array
- *    thrust::device_vector<float> v(4);
- *    v[0] = 1.0f;
- *    v[1] = 2.0f;
- *    v[2] = 3.0f;
- *    v[3] = 4.0f;
- *
- *    float sum_of_squares =
- *     thrust::reduce(thrust::make_transform_iterator(v.begin(), square()),
- *                    thrust::make_transform_iterator(v.end(),   square()));
- *
- *    std::cout << "sum of squares: " << sum_of_squares << std::endl;
- *    return 0;
- *  }
- *  \endcode
- *
- *  The following example illustrates how to use the third template argument to explicitly specify the return type of
- *  the function.
- *
- *  \code
- *  #include <thrust/iterator/transform_iterator.h>
- *  #include <thrust/device_vector.h>
- *
- *  struct square_root
- *  {
- *    __host__ __device__
- *    float operator()(float x) const
- *    {
- *      return sqrtf(x);
- *    }
- *  };
- *
- *  int main()
- *  {
- *    thrust::device_vector<float> v(4);
- *    v[0] = 1.0f;
- *    v[1] = 4.0f;
- *    v[2] = 9.0f;
- *    v[3] = 16.0f;
- *
- *    using FloatIterator = thrust::device_vector<float>::iterator;
- *
- *    // note: float result_type is specified explicitly
- *    thrust::transform_iterator<square_root, FloatIterator, float> iter(v.begin(), square_root());
- *
- *    *iter;   // returns 1.0f
- *    iter[0]; // returns 1.0f;
- *    iter[1]; // returns 2.0f;
- *    iter[2]; // returns 3.0f;
- *    iter[3]; // returns 4.0f;
- *
- *    // iter[4] is an out-of-bounds error
- *  }
- *  \endcode
- *
- *  \see make_transform_iterator
- */
+template <class UnaryFunc, class Iterator>
+struct transform_iterator_reference
+{
+  // by default, dereferencing the iterator yields the same as the function.
+  using type = decltype(::cuda::std::declval<UnaryFunc>()(::cuda::std::declval<iterator_value_t<Iterator>>()));
+};
+
+// for certain function objects, we need to tweak the reference type. Notably, identity functions must decay to values.
+// See the implementation of transform_iterator<...>::dereference() for several comments on why this is necessary.
+template <class Iterator>
+struct transform_iterator_reference<::cuda::std::identity, Iterator>
+{
+  using type = iterator_value_t<Iterator>;
+};
+template <typename Eval, class Iterator>
+struct transform_iterator_reference<functional::actor<Eval>, Iterator>
+{
+  using type = ::cuda::std::remove_reference_t<decltype(::cuda::std::declval<functional::actor<Eval>>()(
+    ::cuda::std::declval<iterator_value_t<Iterator>>()))>;
+};
+
+// Type function to compute the iterator_adaptor instantiation to be used for transform_iterator
+template <class UnaryFunc, class Iterator, class Reference, class Value>
+struct make_transform_iterator_base
+{
+private:
+  using reference  = typename ia_dflt_help<Reference, transform_iterator_reference<UnaryFunc, Iterator>>::type;
+  using value_type = typename ia_dflt_help<Value, ::cuda::std::remove_cvref<reference>>::type;
```

Comment:
Using `::cuda::std::remove_cvref` as a nullary is ... gross

Reply:
What would you prefer instead?